### DISCUSSION TITLE
Hey Maintainers!

### DISCUSSION THREAD

**Discussion by TotalTechGeek** [json-logic/.github#1] [Announcements]
Hey folks; this type of community management is a bit new to me, so I'm open to feedback. 

I've opened a project here to try and track our current goals, which are to help set community guidelines & licensing, and also try to decide on communication approach -
https://github.com/orgs/json-logic/projects/2

This should hopefully not take very long. 

While I generally prefer to communicate over Slack / Discord, where possible I'd like to try to keep these discussions as open and transparent as possible, so I figured GitHub might be an effective forum.

However, these decisions should be driven by the community.

**Comment by TotalTechGeek** [comment]
In the spirit of transparency and openness, I'm also going to link to some of the convos that have taken place on other repos:

A ton of conversation took place here:
- https://github.com/diegoholiveira/jsonlogic/issues/89 (Go)

Some of the other conversations:
- https://github.com/panzi/panzi-json-logic/issues/1 (Python + C)
- https://github.com/Viicos/jsonlogic/issues/34 (Python)
- https://github.com/Open-Payments/datalogic-rs/issues/3 (Rust)
- https://github.com/json-everything/json-everything/issues/820 (C# -- I kinda hijacked an issue to discuss the possibility of working together)

No response (yet):
- https://github.com/allegro/json-logic-kmp/issues/127 (JVM + Swift, written in Kotlin Multiplatform)
- https://github.com/useful-libs/json_logic_ruby/issues/22 (Ruby)

**Comment by codetiger** [comment]
I would like to include naming conventions standardization as part of the governance. When I started with JSONLogic, I found each library having a different notation, like LogicEngine.apply, JSONLogic.apply, JSONLogic.Run etc.

**Reply by TotalTechGeek** [reply]
That might help folks when switching between libraries and ecosystems; so that might be on the table! Though perhaps aliases are allowed for backwards compatibility reasons.

If we go down that route, I think I'd encourage us to separate the library conventions governance from the logic specification

**Reply by codetiger** [reply]
Makes sense, our initial focus should be on spec governance rather than conventions.

**Comment by gregsdennis** [comment]
I think we do need a more informal place to chat.  Things like "I'm at work right now and can't properly respond" doesn't really belong in a GH discussion, but it could still be good information to convey.

**Reply by TotalTechGeek** [reply]
Slack? Discord? Some sort of third party Git IRC?

**Reply by beeme1mr** [reply]
I can create a dedicated channel in the CNCF Slack workspace.

**Reply by TotalTechGeek** [reply]
Let's go ahead and do that for now. 

This is where you can get an invite to the Slack: https://communityinviter.com/apps/cloud-native/cncf

Within the CNCF Slack, you can join the [`#json-logic`channel](https://cloud-native.slack.com/archives/C0849FHVCAV).

**Comment by josephdpurcell** [comment]
PHP FIG is an excellent resource on how to organize specifications and the supporting community: https://www.php-fig.org/. I would be happy to reach out to the group to see if a Q&A could be done if there were well defined questions.

Community driven efforts are a highly complex endeavor, but worthwhile. The impact of FIG extends beyond PHP, for example.

**Comment by Viicos** [comment]
Hello there,

I maintain https://github.com/Viicos/jsonlogic (a Python implementation). I don't know if I'll have much spare time, but I can provide some help from time to time.

I see that there is already some discussion going on to sort out specific details of the implementation (e.g. https://github.com/orgs/json-logic/discussions/10). I think that before doing so, there is some value in:
- discussing how the spec should be worked out? Should this take the form of an RFC [^1]?
- discussing how much the spec should cover.

Regarding the second point, I'm going to introduce briefly some design decisions I made in my implementation.

Whenever possible, I've tried to keep things extensible/abstract:
- I don't define a fixed set of operators, but rather a [registry mechanism](https://jsonlogic.readthedocs.io/latest/api/registry.html) to allow arbitrary operators to be defined and used.
- I don't define a single way to resolve variables, but rather an interface that should be implemented by end users (and still provide two implementations — the JSON Pointer RFC and existing dot syntax). Simplified (this doesn't work as is if you try using my implementation), the interface looks like this:
  ```python
  # I hope you don't mind a basic Python example ;)

  class BaseReferenceParser(ABC):  # This is an abstract class
      @abstractmethod
      def parse(self, reference: str) -> list[str]:
          pass

  # Here is an implementation of the original "dot" format:
  class DotReferenceParser(BaseReferenceParser):
      def parse(self, reference: str) -> list[str]:
          if reference == "":
              return []
          else:
              return reference.split(".")

  from jsonlogic.evaluation import evaluate

  value = evaluate(
      {"var": "some.data"},
      data={"some": {"data": 1}},
      reference_resolver=DotReferenceParser(),
  )
  print(value)
  #> 11
  ```

In other words: whenever possible, I've tried to make things as generic as possible. So what could be done (but of course this is only a suggestion) is to first have a specification describing the general JSON Logic syntax: how operators can be created, how to use them in a JSON Logic expression, etc. We can also define a variable reference parsing/resolving abstract interface, like in the Python example I gave.

A second specification, based on the first one, would be more specific and can/should make opinionated choices when it comes to:
- the set of defined operators, how they behave with different data types (e.g. should we allow `<`/`>`/`<=`/`>=` on strings?).
- the default variable reference format.

---

Additional note: I my implementation, I kind of already have a distinction between the two previous paragraphs. Given a raw JSON Logic expression, I first parse it into a `JSONLogicExpression`:

```python
expr = JSONLogicExpression.from_json(
    {
        "*": [
            {"var": "some_var"},
            2
        ]
    }
)
print(expr)
>>> JSONLogicExpression(
>>>     expression={'*': [
>>>             JSONLogicExpression(expression={'var': ['some_var']}),
>>>             2,
>>>         ]
>>>     }
>>> )
```

The parsing doesn't do much, but notice that it makes sure single operand operators (like `"var"`) have their arguments wrapped in a list if necessary. However, no operator definition is involved here at this stage.

[^1]: Drafting an RFC is a long process, so might not be worth going this way. But at least the [RFC redacting guidelines](https://tools.ietf.org/html/rfc7322) could be used.

**Reply by TotalTechGeek** [reply]
> discussing how the spec should be worked out? Should this take the form of an RFC [1](https://github.com/orgs/json-logic/discussions/1#user-content-fn-1-33caa11ab1a2fe5a9aae5601423eca7d)?

The eventual goal is to create an RFC and turn this into a standard; 

But for expediting the communication process and getting implementations to be compatible, we'll begin the initiative with test specifications, and creating a robust standard test suite.

A few standard operators may also be introduced in this process. 

This is sort of a JSON Logic v2 + standardization initiative, but I'd like to retain backwards compatibility with current mainline operators to avoid invalidating most implementations and existing logic in the wild.

> discussing how much the spec should cover

Yep! 

This is up for discussion and one of the goals of the initiative. 

I'd really like to resolve a lot of common inconsistencies between implementations and common issues that users have experienced while leveraging JSON Logic. And ideally, we'd also make it more generally capable as an AST.

> the set of defined operators, how they behave with different data types (e.g. should we allow </>/<=/>= on strings?)

For example, I think this should be included in the specification in the first cut of this initiative in the test-spec phase. I actually think most implementations already implement this despite not being in the standard test suite.

**Reply by gregsdennis** [reply]
Regarding having two separate specifications, I advise following the RFC9535 JSON Path example: Specification with registry.  The specification defines the syntax, including an extension point: expression functions.  Then it defines a base set of functions.  The only separate piece is an official registry for specifying new functions.  I think this is a good approach for JSON Logic.

With JSON Schema, we have two separate specification documents: Core and Validation.  The idea was two-fold.  First, the spec was becoming absurdly long.

And second, Core ideally contains the minimum viable set of functionality to make a "JSON Schema"-like thing work.  Validation contains additional keywords that actually make it useful.  (This is similar to what you're proposing.)  The idea was to allow additional use cases where validation wasn't as important (e.g. code generation) to define their own extensions on top of Core.  I don't think that we got everything worked out, and we're reconsidering some aspects of it for the next release.

**Reply by gregsdennis** [reply]
I also like the idea of supporting both the dot-syntax and JSON Pointer, at least in initial versions of the spec.  If JSON Pointer is something that you want to migrate to, you can specify that implementations support both and produce a warning of some sort when the dot-syntax is used.

(In my implementation, I actually just convert the dot syntax to JSON Pointer, then evaluate using _JsonPointer.Net_, which is also mine.)

**Reply by TotalTechGeek** [reply]
As noted in the other discussion, I'd like to be very cautious about any changes that break backwards compatibility / could fragment the JSON Logic ecosystem via misc configuration flags, especially within the specification.

I'd rather: 
- Perform a slight modification to improve access for the current format*
**and**
- Get rid of path parsing semantics **entirely** by introducing a `val` operator, and encourage folks to use that instead. (No dot syntax, no json pointer, making things entirely AST-like)

If we're aiming to change out functionality, I think I'd rather introduce a new operator and get it well-supported, instead of invalidating every existing implementation & creating ecosystem forks 😅 

*Perhaps*, we deprecate a few of the operators with frustrating behaviors (`var`, `missing`, etc), while documenting their behaviors, and continue testing them for now, while encouraging a switch to the new operators.

---

Just adding a stray thought. One thing that's potentially nice about `val` is that you could potentially chain it, `{ val: { rfc6901: '/foo/bar' } }`


\* And honestly I could skip this step. Keep `var` simple, go all-in on `val` to get rid of path semantics in a string

**Reply by rmannibucau** [reply]
Think we should step back and think to usage cases. For me there are two:

1. Programmatic -> there anything proposed works
2. Static/interop between layer (often backend/frontend) or configuration -> there being completely an ast without any inline programming capability is a strength since it enables very easy validation, tooling and harnessing

So overall, if jsonpath or a custom inline ds is going to be a thing, I'd vote to have a minimal set where it is not supported and an extension where it can be supported by impl.
Some spec umbrella use the notion of "profile" so we could get a "static profile" and "dynamic profile" maybe.

Then an impl can be compliant to a single profile if desired and consumers can pick the requirement(s) they need (as we all do when picking a dependency).

**Reply by TotalTechGeek** [reply]
> I'd vote to have a minimal set where it is not supported

This minimal set will be larger than you'd like, but I actually do agree to an extent here. 😄

The base will include existing operators (with a few of them being "deprecated"), and potentially a few operators that resolve some common asks (embedding data). It **will** end up including var, potentially deprecated. 

However, if folks agree to `val`, which removes path parsing semantics, that'll be available for folks to switch to.

For example, the `rfc6901` operator I mentioned above shouldn't be in the minimal set, it would be an extension. 

> Static/interop between layer
> "profiles"

This is alright! I'm on board w/ enabling this / making it easier to use JSON Logic as an AST.

In `json-logic-engine`, I have an export called `asLogicSync`. So if I wanted to [serialize password rules](https://github.com/TotalTechGeek/simple-password-rules), I could call

```javascript
import * as rules from 'simple-password-rules'

// Turns all of the functions into serializable logic. 
// Does not keep the other operators; I have to opt into it by passing in a keep array.
const builder = asLogicSync(rules) 

const check = builder({
  password: [
    { min: 8 },
    { max: 16 },
    { hasDigit: 1 },
    { hasSpecial: 1 }
  ]
})

const errors = check('password1')
if (errors) {
    throw new Error(errors) //  The password provided does not contain enough "special" characters. Needs 1 more. 
}
```

Alternatively, you could create a LogicEngine in the same way `new LogicEngine(rules)`, basically just substituting out the supported operators.

I've even used JSON Logic to generate Mongo and SQL Queries, and substituted each operator so that the same JSON Logic could run in different environments. (In the Application, in the DB, etc)

Anyway -- this seems more a matter of communicating in each library:  "Here's how you opt-out of the built in operators to use your own"

*Perhaps* we could suggest conventions later on w/r/t the base and extensions, to make it easy for folks to do something like:
`new LogicEngine({ ...base, ...certLogic })`

We can, however, for the purposes of keeping track of this type of user, have two sections of the specification, one to explain evaluation, one describing each operator.

**Reply by rmannibucau** [reply]
Think we should define the minimal profile as not supporting any backward compat.
This is key if you target any portability and being consumed easily and potentially not break their database or configuration if they switch of language/technology (a common example is regex extensions are highly likely to not be portable either in terms of impl or runtime/env for good reasons).

So more I'm thinking to it more I think having a kind of shared spec makes sense if we can reach this minimal definition - without preventing extensions, I think a key based extension mechanism is great for ex - but this is what people will know they can rely on in complex env and only what they can rely on IMHO (my XP on that was a backend - java - and frontend - js - share of the rules, some of the syntax and behavior was not nicely portable as of today for ex and this part was not strictly needed but since it was there frontend used it and broke the backend without seeing it as quickly as it should have been).

**Reply by TotalTechGeek** [reply]
### Backwards Compatibility with `tests.json` is non-negotiable

> Think we should define the minimal profile as not supporting any backward compat.

I'm sorry, but this is not a position that will be budged on.  

There are a minimum of [19 implementations](https://github.com/orgs/json-logic/discussions/14#discussion-7673600) that have agreed to be involved in / associated with this project, each with their own user-bases and logic that is serialized / transmitted somewhere.

Documentation has been created, users have adopted it, organizations use this to power their applications. 

We will not be breaking backwards compatibility or redefining any operators in such a way where their behavior will increase ecosystem fragmentation and kill off the main goal of JSON Logic unification. 

I am okay with defining new replacement operators, and deprecating some of the old operators, but any major obstacle that prevents users and applications from upgrading will not be acceptable.

I also believe that a break in backwards compatibility would lead to a lack of credibility in the initiative, and cause future adopters to have extreme distrust in JSON Logic. (I really wouldn't feel comfortable if I felt like my logic wouldn't run in a couple years, and really wouldn't end up being portable across implementations. I'd put more faith in a solution that tries to respect backwards compat.)

[I've made it clear that I'm receptive to input and explore options; yours in particular. For example, I agree that implementing an AST-like way to access variables is a solid recommendation. I'm in favor of encouraging it as the main choice going forward!](https://github.com/orgs/json-logic/discussions/10#discussioncomment-11502029)

[OpenFeature](https://github.com/open-feature), for example, uses JSON Logic. 

@beeme1mr / @toddbaert might have thoughts, but breaking backwards compatibility would not only harm their project, but their users, as their existing configurations for targeting rules could potentially become invalidated. 

This would serve as a major roadblock for them upgrading; and potentially other organizations too.

### However

I am, however, on the same page with regards to extensions and not overloading the core spec with these extensions. 

Regex operators, Date extensions, etc.


Those are all good candidates for extensions to the JSON Logic spec. I'm also on the same page on trying to improve JSON Logic as a general AST. 

(I do, however, want the compat tables to show compliance with these common extensions, so folks know what features they can comfortably use.)

**Reply by rmannibucau** [reply]
> We will not be breaking backwards compatibility or redefining any operators in such a way where their behavior will increase ecosystem fragmentation and kill off the main goal of JSON Logic unification.

As mentionned, the fact to break or not is up to the vendors, this part is not a topic.
For me the only topic is what will be the future, a stack on legacy lack of agreements or a shared agreements.

> I also believe that a break in backwards compatibility would lead to a lack of credibility in the initiative, and cause future adopters to have extreme distrust in JSON Logic.

Guess we are in "believe" land so not sure it is worth discussing but from what I saw in JavaEE/JakartaEE land, breaking can just not change the users. In another ecosystem, JSON-Schema is a good example you can break and still be adopted all the time - not saying impl must not think to backward compat but that it is not important for a v1 of a spec.

**Reply by TotalTechGeek** [reply]
> for me the only topic is what will be the future, a stack on legacy lack of agreements or a shared agreements.

I believe that most of the community will agree on most of the existing operators; becoming a shared agreement. 

I believe that `var`, `missing` and `missing_some` might be the more controversial operators because they're more difficult to retrofit, because they have additional pathing syntax baked into them (and several arguments), and are less "AST-like".

We can address this easily by describing these as legacy, deprecated operators that we expect implementations to continue supporting for the foreseeable future, and encourage folks to use `val` or some replacement operator that the community does agree upon. 

This also gives plenty of time for folks to implement a migration path, if we inevitably do remove some of these operators.

Defining new operators for some of these controversial behaviors from the original is a much cleaner path than redefining an existing one. It also makes implementing a migration easier.

With this path, we can even take the path of labeling `var`, `missing`, etc. as a "legacy extension" 

But I don't want to _break_ `var` for users. 

---

There's room to discuss whether `var` and such should be left out of the RFC, if we take the path of defining replacement operators.

But for the compatibility suite, I think it's necessary.

**Reply by rmannibucau** [reply]
> But I don't want to break var for users.

We agree on that for the migration time but where we are not yet aligned I guess is that the spec needs to mention anythingl egacy. Understand me, the spec doesn't exist and would start by defining legacy operators, this is starting with debt and I don't think it helps the goal since users will start looking at already deprecated stuff so I think it is worth not mentionning it in the spec - but having a dedicated page on implementation vendors or a doc page explaining how to migrate in the spec makes sense to me, but no assumption it must be supported by a vendor.

So I'm clearly in favor of not trying to be strictly  backward compatible with something which doesn't exist (previous version of the spec) but documentation what had been the replacements (like JPA explains what replaces hibernates specific annotations for ex).

**Reply by TotalTechGeek** [reply]
>  is that the spec needs to mention anything legacy

I'm potentially okay with leaving `var` (and similar) out of the RFC / formal specification for core JSON Logic.

I'm also okay with defining those other operators as a legacy "extension". We would test all accepted community extensions in our compatibility tables, and make it easy for folks to look up what operators are generally well supported, similar to ["Can I Use"](https://caniuse.com/)

The extensions would be their own specs. 

Once we solidify things with the base, I'd like for this community to be a channel for folks to propose common community extensions & use this as a hub for helping developers with resources needed to get the extensions implemented.

**Reply by rmannibucau** [reply]
I like the caniuse idea using the spec and the original impl as comparison points :+1:

**Reply by gregsdennis** [reply]
When we were writing RFC9535, our primary problem was implementation alignment.  Implementations only had a [blog post](https://goessner.net/articles/JsonPath/) and a reference implementation as their source material.  As a result, many implementations kinda did their own thing.  We needed the JSON Path Comparison Project to see how diverse the implementations really were.  Backward incompatibility was inevitable for that project, despite our efforts to minimize it.

I think the JSON Logic implementations should be significantly more aligned given the (more, yet still incomplete) documentation, at least two reference implementations, and a moderate (though also incomplete) test suite.  It's reasonable that we can find a large compatible intersection of support.

**Comment by dslmeinte** [comment]
For the [specification of CertLogic](https://github.com/ehn-dcc-development/eu-dcc-business-rules/tree/main/certlogic/specification) I didn't use a specific, pre-existing format.

One advice: it's probably better to first specify the basics and 1 or 2 operators completely to bits (including defining a good test suite), than it is to try and specify everything including all operators all at once.

**Reply by TotalTechGeek** [reply]
I *think* the strategy for now is:

1. Develop a rigorous test suite, and make test specs a big part of the proposal process, (especially since it's all JSON). [Our test format is documented here](https://github.com/json-logic/.github/blob/main/TEST_FORMAT.md). This will serve as a pseudo-specification for the purposes of moving quickly and improving compatibility.

2. Once we have a general set of operators that we're comfortable with and are well-defined, and we've addressed a handful of common issues users face either adopting or using JSON Logic, and we've improved ecosystem compatibility, we'll work on upgrading our test specification to a more formal RFC. This will likely take months.

While we're working on the RFC, we may use your specification as a reference for working up from the test spec to something a bit more formal. We'll definitely need to be rigorous with how we define each operator as an "Atom" (as you've described it)

**Reply by JTeeuwissen** [reply]
Before developing a rigorous test suite, might it be beneficial to draft and discuss some ground rules for json logic?

For example:
Which operators are included and which ones are not. (E.g. do we really want to use js equality? And what is the name for ternary?)
How do these operators function (e.g. should + work on more than two arguments?)
And how to deal with invalid input (e.g. more parameters than allowed, do we ignore them?)

Some decisions might make sense for the js implementation, but should be reconsidered in this wider ecosystem.

**Reply by rmannibucau** [reply]
Agree. Ideally we could get multiple flavors, a) no-operator defined but how it works (useful for product integration where the security - wide sense - is key and operators must be strictly controlled and potentially their behavior changed), b) minimal set (built-in safe operators), c) maybe a core set (common operators), d) (unspecified) implementation/language specific.

I fear we'l lneed the falsy/truthy impl - js style or not - cause otherwise it becomes very verbose to use it and makes it harder to integrate and welcome new users but what is important is the extensibility - custom operators - will stay a key feature of all impl even if it can't be strictly specified outside a language/API.

Hope it makes sense.

**Reply by JTeeuwissen** [reply]
Personally I have never felt the need for truthy values or equality type conversion, an explicit conversion seems easier to understand. But it's a great example of something that should be discussed before writing a spec.

**Reply by rmannibucau** [reply]
Agree, maybe you can open a ticket/discussion about it?

side note: i'm interested in `an explicit conversion seems easier to understand`, so you prefer to add a wrapper in json logic to test null or lengths? I can see cases it is better, fully agree on that but also most cases where it makes the json more verbose for almost no gain so I'm not 100% sure. Maybe it could be some shortcut in "c" but not in "a" and "b" - thinking out loud.

**Reply by TotalTechGeek** [reply]
Hey there; there are some closed discussions as well as this guy to kick off ground rules -
https://github.com/orgs/json-logic/discussions/16

For now - 
JSON Logic "Core" is the end goal of this initiative; and compatibility with the current tests.json is required for core*

\* at least in the sense that any changes made to Core do not break tests.json

We may end up deprecating a few operators to be considered legacy extensions, and not part of "Core".

I think we'll end up making good on "a", "b" and "c" as described by @rmannibucau, 

Where "b" is JSON Logic Core and "c" is Core + Community extensions.

"a" would just be the AST specification, and users would define their own operators and parsing rules. 

This means of course, folks could change how truthiness works for their app, but it would diverge from core.

-- 

How we handle invalid input is within scope of the discussions, also variadic nature of the comparison operators.

The TC will help organize a lot of these thoughts into a consolidated document
